auth/auth.go

```diff
@@ -0,0 +1,281 @@
+package auth
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/coreos/go-oidc"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+	"golang.org/x/oauth2"
+
+	"github.com/sourcegraph/sourcegraph-accounts-sdk-go/scopes"
+)
+
+// Config contains the configuration for the SAMS authentication handler.
+type Config struct {
+	// Issuer is the SAMS instance URL, e.g. "https://accounts.sourcegraph.com".
+	Issuer string
+	// ClientID is the SAMS client ID, e.g. "sams_cid_xxx".
+	ClientID string
+	// ClientSecret is the SAMS client secret, e.g. "sams_cs_xxx".
+	ClientSecret string
+	// RequestScopes is the list of requested scopes for access tokens that are
+	// issued to this client.
+	RequestScopes []scopes.Scope
+	// RedirectURI is the URL to redirect to after the user has authenticated.
+	RedirectURI string
+
+	// FailureHandler is the HTTP handler to call when an error occurs. Use
+	// ErrorFromContext to extract the error.
+	FailureHandler http.Handler
+	// StateSetter sets the randomly-generated state to the per-user session.
+	StateSetter func(w http.ResponseWriter, r *http.Request, state string) error
+	// StateGetter gets the state from the per-user session.
+	StateGetter func(r *http.Request) (string, error)
+	// StateDeleter deletes the state from the per-user session.
+	StateDeleter func(w http.ResponseWriter, r *http.Request)
+}
+
+// Error is an error that occurred during the authentication process.
+type Error struct {
+	// StatusCode is the HTTP status code to respond with.
+	StatusCode int
+	// Cause is the error that caused the failure.
+	Cause error
+}
```

Comment:
This is fine, but just for my own understanding, to make this a "proper" Golang error you would need to implement `func (e *Error) Error() string` and `func (e *Error) Unwrap() error` right?

Reply:
> `func (e *Error) Error() string` and `func (e *Error) Unwrap() error` right?

Only `func (e *Error) Error() string` is required to make it an `error`, `func (e *Error) Unwrap() error` is optional that, if you are not wrapping any error, you don't need to (or if you want to intentionally hide the original error).